src/orange/surfaces/GeneralQuadric.hh

```diff
@@ -0,0 +1,206 @@
+ */
+CELER_FUNCTION GeneralQuadric::GeneralQuadric(const Real3& abc,
+                                              const Real3& def,
+                                              const Real3& ghi,
+                                              real_type    j)
+    : a_(abc[0])
+    , b_(abc[1])
+    , c_(abc[2])
+    , d_(def[0])
+    , e_(def[1])
+    , f_(def[2])
+    , g_(ghi[0])
+    , h_(ghi[1])
+    , i_(ghi[2])
+    , j_(j)
+{
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from raw data.
+ */
+CELER_FUNCTION GeneralQuadric::GeneralQuadric(Storage data)
+    : a_(data[0])
+    , b_(data[1])
+    , c_(data[2])
+    , d_(data[3])
+    , e_(data[4])
+    , f_(data[5])
+    , g_(data[6])
+    , h_(data[7])
+    , i_(data[8])
+    , j_(data[9])
+{
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Determine the sense of the position relative to this surface.
+ */
+CELER_FUNCTION SignedSense GeneralQuadric::calc_sense(const Real3& pos) const
+{
+    const real_type x = pos[0];
+    const real_type y = pos[1];
+    const real_type z = pos[2];
+
+    real_type result = (a_ * x + d_ * y + f_ * z + g_) * x
+                       + (b_ * y + e_ * z + h_) * y + (c_ * z + i_) * z + j_;
+
+    return real_to_sense(result);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate all possible straight-line intersections with this surface.
+ */
+CELER_FUNCTION auto
+GeneralQuadric::calc_intersections(const Real3& pos,
+                                   const Real3& dir,
+                                   SurfaceState on_surface) const
+    -> Intersections
+{
+    const real_type x = pos[0];
+    const real_type y = pos[1];
+    const real_type z = pos[2];
+    const real_type u = dir[0];
+    const real_type v = dir[1];
+    const real_type w = dir[2];
+
+    // Quadratic values
+    real_type a = (a_ * u + d_ * v) * u + (b_ * v + e_ * w) * v
+                  + (c_ * w + f_ * u) * w;
+    real_type b = (2 * a_ * x + d_ * y + f_ * z + g_) * u
+                  + (2 * b_ * y + d_ * x + e_ * z + h_) * v
+                  + (2 * c_ * z + e_ * y + f_ * x + i_) * w;
+    real_type c = ((a_ * x + d_ * y + g_) * x + (b_ * y + e_ * z + h_) * y
+                   + (c_ * z + f_ * x + i_) * z + j_);
+
+    return detail::QuadraticSolver::solve_general(a, b / 2, c, on_surface);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate outward normal at a position.
+ */
+CELER_FUNCTION Real3 GeneralQuadric::calc_normal(const Real3& pos) const
+{
+    const real_type x = pos[0];
+    const real_type y = pos[1];
+    const real_type z = pos[2];
+
+    Real3 norm;
+    norm[0] = 2 * a_ * x + d_ * y + f_ * z + g_;
+    norm[1] = 2 * b_ * y + d_ * x + e_ * z + h_;
+    norm[2] = 2 * c_ * z + e_ * y + f_ * x + i_;
+
+    normalize_direction(&norm);
+    return norm;
```

Comment:
Hi @sethrj, while looking around in the ORANGE code, I came across this method and how you use it to calculate the safety, ie the smallest distance from a point to the surface. Do you have (public) notes how this is derived mathematically?

I'm wondering about the cases of ellipsoids (or the special case of spheres), cones and cylinders where there are only quadratic terms. If I read the code correctly, it would then produce a null vector for a point at the origin. This likely will give no meaningful answer for the safety, which is trying to intersect a line following the normal vector, will it?

Reply:
@hahnjo The normal vector at a point on a surface is just the [gradient of the surface equation](https://docs.openmc.org/en/latest/methods/geometry.html#id7) evaluated at that point (x,y,z). I'm not sure what meaning it holds if you evaluate it for a point that's _not_ on the surface (as it appears to be used for the safety distance) so maybe @sethrj can enlighten us both :smile:

Reply:
No, I'm fairly convinced this doesn't hold mathematically. I've played with [tubes on GeoGebra](https://www.geogebra.org/calculator/v9srqvrs) and it works for `a = b = 1`. For `a = 2` you can already see that the safety sphere intrudes into the tube, and for `a = 3` the method doesn't find any intersection at all.

Reply:
Well, you're definitely right! Sorry for the misleading class and incorrect algorithm. Maybe the safety could be approximated with a few iterations with steepest descent?

Thanks for digging into this, and for the link to the very nifty visualization tool!

Reply:
I created an issue #352 so we can have further discussions there rather than buried away inside this code review comment 😄

Reply:
>  What we need for safety is a sort of "conservative" estimate, ie the returned value can be too low but it must never be too large, I think

Exactly. Anything that can give us a conservative estimate (especially if the accuracy increases near the surface itself) would be good.